client/lib/domains/dns/reducer.js

```diff
@@ -46,6 +44,36 @@ function deleteDns( state, domainName, record ) {
 	return update( state, command );
 }
 
+function markDnsForDeletion( state, domainName, record, isBeingDeleted ) {
+	const index = findDns( state, domainName, record ),
+		updatedRecord = Object.assign( {}, record, {
+			isBeingDeleted,
+		} );
+
+	if ( index === -1 ) {
+		return state;
+	}
+
+	const command = {
+		[ domainName ]: {
+			records: {
+				[ index ]: {
+					$merge: updatedRecord
+				}
+			}
+		}
+	};
+
+	return update( state, command );
+}
+
+function findDns( state, domainName, record ) {
+	const { id, data, name, type } = record,
+		matchingFields = isUndefined( id ) ? { data, name, type } : { id, data, name, type };
+
+	return findIndex( state[ domainName ].records, matchingFields );
+}
+
```

Comment:
It seems you don't need `state` or `domainName` in this function. You can just pass in `records`.

    function findDnsIndex( records, { id, data, name, type } ) {
      const fields = isUndefined( id ) ? { data, name, type } : { id, data, name, type };
      return findIndex( records, matchingFields );
    }